backends/ref/ceed-ref-operator.c

```diff
@@ -701,10 +737,32 @@ static int CeedOperatorSetupFieldsAtPoints_Ref(CeedQFunction qf, CeedOperator op
         CeedCallBackend(CeedOperatorFieldGetElemRestriction(op_fields[j], &rstr_j));
         if (vec_i == vec_j && rstr_i == rstr_j) {
           CeedCallBackend(CeedVectorReferenceCopy(e_vecs[i], &e_vecs[j]));
+          CeedCallBackend(CeedVectorReferenceCopy(e_vecs_full[i + start_e], &e_vecs_full[j + start_e]));
           skip_rstr[j] = true;
         }
       }
     }
+  } else {
+    for (CeedInt i = num_fields - 1; i >= 0; i--) {
+      CeedVector          vec_i;
+      CeedElemRestriction rstr_i;
+
+      CeedCallBackend(CeedOperatorFieldGetVector(op_fields[i], &vec_i));
+      CeedCallBackend(CeedOperatorFieldGetElemRestriction(op_fields[i], &rstr_i));
+      for (CeedInt j = i - 1; j >= 0; j--) {
+        CeedVector          vec_j;
+        CeedElemRestriction rstr_j;
+
+        CeedCallBackend(CeedOperatorFieldGetVector(op_fields[j], &vec_j));
+        CeedCallBackend(CeedOperatorFieldGetElemRestriction(op_fields[j], &rstr_j));
+        if (vec_i == vec_j && rstr_i == rstr_j) {
+          CeedCallBackend(CeedVectorReferenceCopy(e_vecs[i], &e_vecs[j]));
+          CeedCallBackend(CeedVectorReferenceCopy(e_vecs_full[i + start_e], &e_vecs_full[j + start_e]));
+          skip_rstr[j]       = true;
+          apply_add_basis[i] = true;
+        }
+      }
+    }
```

Comment:
I think this logic is repeated enough to package up into it's own function that all the backends can access? Maybe some abstraction of it (ie. just report back the indices of fields with matching restrictions and vectors).

Reply:
I'm struggling to come up with a good helper function because the input and output loops run in opposite directions from each other. Essentially I need the first instance of the vec/rstr pair and the last instance of, and then I'd still need to loop over those arrays to do the reference copying and mark the rstr to skip or the basis to apply-add. I don't have a signature of such a helper that I actually like

Reply:
I think I'll merge as is for now and circle back to helper if I end up using this approach this in Gen as well